File: token-flamegraph/terminal.py
"""
Terminal-based rendering of the token visualization dashboard.

Renders directly in the terminal with ANSI colors, Unicode bar charts,
and structured layout — no browser needed.
"""

import shutil
from flamegraph import SessionViz, session_to_viz, ToolSpan
from parser import Session


# ANSI color codes
COLORS = {
    "thinking": "\033[91m",   # red
    "text":     "\033[94m",   # blue
    "read":     "\033[92m",   # green
    "write":    "\033[93m",   # yellow
    "bash":     "\033[95m",   # magenta
    "search":   "\033[32m",   # dark green
    "agent":    "\033[31m",   # dark red
    "other":    "\033[96m",   # cyan
    "cache":    "\033[90m",   # gray
    "fresh":    "\033[37m",   # light gray
}
RESET = "\033[0m"
BOLD = "\033[1m"
DIM = "\033[2m"

# Bar characters
FULL_BLOCK = "█"
LIGHT_BLOCK = "░"
MED_BLOCK = "▒"
DARK_BLOCK = "▓"


def _c(category: str, text: str) -> str:
    """Colorize text by category."""
    color = COLORS.get(category, "")
    return f"{color}{text}{RESET}" if color else text


def _bar(value: int, max_val: int, category: str, width: int) -> str:
    """Render a colored bar."""
    if max_val == 0:
        return ""
    filled = max(0, int(value / max_val * width))
    return _c(category, FULL_BLOCK * filled)


def _fmt(n: int) -> str:
    """Format number with commas."""
    return f"{n:,}"


def _pct(n: int, total: int) -> str:
    if total == 0:
        return "0%"
    return f"{n * 100 / total:.1f}%"


def render_terminal(session: Session, title: str = "Token Flamegraph") -> str:
    """Render the full dashboard as a string with ANSI colors."""
    viz = session_to_viz(session)
    term_width = min(shutil.get_terminal_size().columns, 120)
    bar_width = max(30, term_width - 50)
    lines = []

    def ln(s=""):
        lines.append(s)

    # Header
    ln(f"{BOLD}{'═' * term_width}{RESET}")
    ln(f"{BOLD}  {title}{RESET}")
    if viz.model:
        ln(f"  {DIM}Model: {viz.model}{RESET}")
    ln(f"{BOLD}{'═' * term_width}{RESET}")

    # Stats
    cache_pct = _pct(viz.total_cache, viz.total_input)
    active_turns = [t for t in viz.turns if t.output_tokens > 0]
    avg_score = sum(t.efficiency_score for t in active_turns) / len(active_turns) if active_turns else 0
    if avg_score >= 70:
        score_color = "\033[92m"
    elif avg_score >= 40:
        score_color = "\033[93m"
    else:
        score_color = "\033[91m"

    ln()
    ln(f"  {BOLD}Output:{RESET} {_fmt(viz.total_output)}    "
       f"{BOLD}Thinking:{RESET} {_fmt(viz.total_thinking)}    "
       f"{BOLD}Tool Calls:{RESET} {viz.num_tool_calls}    "
       f"{BOLD}Turns:{RESET} {len(active_turns)}    "
       f"{BOLD}Score:{RESET} {score_color}⚡{avg_score:.0f}{RESET}")
    # Cost estimation (Anthropic pricing per MTok)
    # Opus: input $3, cached input $0.30, output $15
    # Sonnet: input $3, cached input $0.30, output $15
    fresh_input = viz.total_input - viz.total_cache
    cost_input = (fresh_input / 1_000_000) * 3.0
    cost_cached = (viz.total_cache / 1_000_000) * 0.30
    cost_output = (viz.total_output / 1_000_000) * 15.0
    cost_total = cost_input + cost_cached + cost_output
    cost_no_cache = ((viz.total_input / 1_000_000) * 3.0) + cost_output
    saved = cost_no_cache - cost_total

    ln(f"  {BOLD}Input:{RESET}  {_fmt(viz.total_input)}    "
       f"{BOLD}Cached:{RESET} {_fmt(viz.total_cache)} ({cache_pct})    "
       f"{BOLD}Cost:{RESET} ${cost_total:.2f} (saved ${saved:.2f} from cache)")

    # Timeline waterfall
    ln()
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")
    ln(f"  {BOLD}TIMELINE — Output Tokens per Turn{RESET}")
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")

    max_out = max((t.output_tokens for t in viz.turns), default=1)

    for turn in viz.turns:
        if turn.output_tokens == 0:
            continue

        # Build colored bar segments
        segments = []
        if turn.thinking_tokens > 0:
            segments.append(("thinking", turn.thinking_tokens))
        for ts in turn.tool_spans:
            segments.append((ts.category, ts.total_tokens()))
        text_tok = max(0, turn.output_tokens - turn.thinking_tokens - turn.tool_tokens)
        if text_tok > 0:
            segments.append(("text", text_tok))

        # Render multi-colored bar
        bar_str = ""
        for cat, tok in segments:
            w = max(1, int(tok / max_out * bar_width)) if tok > 0 else 0
            bar_str += _c(cat, FULL_BLOCK * w)

        # Efficiency score with color
        score = turn.efficiency_score
        if score >= 70:
            score_color = "\033[92m"  # green
        elif score >= 40:
            score_color = "\033[93m"  # yellow
        else:
            score_color = "\033[91m"  # red
        score_str = f"{score_color}{score:4.0f}{RESET}"

        ln()
        ln(f"  {BOLD}Turn {turn.index:>2d}{RESET} │{bar_str}│ {_fmt(turn.output_tokens):>7s} tok  ⚡{score_str}")

        # Tool list
        if turn.tool_spans:
            tool_names = []
            for ts in turn.tool_spans:
                short_name = ts.name
                if len(short_name) > 30:
                    short_name = short_name[:28] + ".."
                tool_names.append(_c(ts.category, short_name))
            # Show up to 6 tools
            shown = tool_names[:6]
            extra = len(tool_names) - 6
            tool_line = ", ".join(shown)
            if extra > 0:
                tool_line += f" {DIM}+{extra} more{RESET}"
            ln(f"         │ {tool_line}")

        # Sub-agent children
        for ts in turn.tool_spans:
            if ts.children:
                child_names = [_c(c.category, c.name) for c in ts.children[:6]]
                more = f" {DIM}+{len(ts.children) - 6} more{RESET}" if len(ts.children) > 6 else ""
                agent_name = ts.name[:40]
                ln(f"         │  └─ {_c('agent', agent_name)}")
                ln(f"         │     {', '.join(child_names)}{more}")

    # Output breakdown
    ln()
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")
    ln(f"  {BOLD}OUTPUT TOKEN BREAKDOWN{RESET}")
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")

    categories = [
        ("Thinking", viz.total_thinking, "thinking"),
        ("Text output", viz.total_text, "text"),
    ]
    cat_names = {
        "read": "File reads", "write": "File writes", "bash": "Shell commands",
        "search": "Search", "agent": "Sub-agents", "other": "Other tools"
    }
    for cat, tokens in sorted(viz.tool_category_tokens.items(), key=lambda x: -x[1]):
        categories.append((cat_names.get(cat, cat), tokens, cat))

    total_cat = sum(c[1] for c in categories)
    max_cat = max(c[1] for c in categories) if categories else 1

    for name, tokens, cat in categories:
        b = _bar(tokens, max_cat, cat, 35)
        pct = _pct(tokens, total_cat)
        ln(f"  {name:<16s} {b} {_fmt(tokens):>7s} ({pct:>5s})")

    # Wall time per turn (only when duration data is available, e.g. teleport exports)
    # Contributed by session 01AjHxHw
    if any(t.duration_ms > 0 for t in viz.turns):
        ln()
        ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")
        ln(f"  {BOLD}WALL TIME PER TURN{RESET}")
        ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")

        max_dur = max((t.duration_ms for t in viz.turns), default=1)
        total_wall = sum(t.duration_ms for t in viz.turns)
        total_api = sum(t.duration_api_ms for t in viz.turns)
        total_overhead = total_wall - total_api

        for turn in viz.turns:
            if turn.duration_ms == 0:
                continue
            api_ms = turn.duration_api_ms
            overhead_ms = turn.duration_ms - api_ms
            api_bar = _bar(api_ms, max_dur, "thinking", 30)
            overhead_bar = _bar(overhead_ms, max_dur, "bash", 10)
            model_pct = api_ms * 100 // max(turn.duration_ms, 1)
            secs = turn.duration_ms / 1000
            ln(f"  T{turn.index:<3d} {api_bar}{overhead_bar} {secs:>6.1f}s ({model_pct}% model)")

        wall_s = total_wall / 1000
        api_s = total_api / 1000
        over_s = total_overhead / 1000
        ln(f"  {DIM}Total: {wall_s:.1f}s wall / {api_s:.1f}s model / {over_s:.1f}s overhead{RESET}")

    # Context growth
    ln()
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")
    ln(f"  {BOLD}INPUT COST PER TURN{RESET}")
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")

    max_input = max((t.input_tokens for t in viz.turns), default=1)
    for turn in viz.turns:
        if turn.input_tokens == 0:
            continue
        cache_pct_val = turn.cache_read_tokens * 100 // max(turn.input_tokens, 1)
        cache_bar = _bar(turn.cache_read_tokens, max_input, "cache", 40)
        fresh_bar = _bar(turn.fresh_input_tokens, max_input, "fresh", 40)
        ln(f"  T{turn.index:<3d} {cache_bar}{fresh_bar} {_fmt(turn.input_tokens):>7s} ({cache_pct_val}% {DIM}cached{RESET})")

    # Top tools
    ln()
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")
    ln(f"  {BOLD}TOP TOKEN CONSUMERS{RESET}")
    ln(f"  {DIM}{'─' * (term_width - 4)}{RESET}")

    if viz.top_tools:
        max_tool = viz.top_tools[0][1]
        for name, tokens in viz.top_tools[:10]:
            short = name[:42]
            # Detect category from name
            cat = "other"
            if name.startswith("Read"): cat = "read"
            elif name.startswith("Edit") or name.startswith("Write"): cat = "write"
            elif name.startswith("Bash"): cat = "bash"
            elif name.startswith("Grep") or name.startswith("Glob"): cat = "search"
            elif name.startswith("Agent"): cat = "agent"
            b = _bar(tokens, max_tool, cat, 30)
            ln(f"  {short:<44s} {b} {_fmt(tokens):>6s}")

    ln()
    ln(f"{BOLD}{'═' * term_width}{RESET}")

    # Legend
    legend_items = [
        ("thinking", "Thinking"), ("text", "Text"), ("write", "Write"),
        ("read", "Read"), ("bash", "Shell"), ("search", "Search"), ("agent", "Agent"),
    ]
    legend = "  ".join(f"{_c(cat, FULL_BLOCK * 2)} {name}" for cat, name in legend_items)
    ln(f"  {legend}")
    ln()

    return "\n".join(lines)

File: token-flamegraph-rs/src/terminal.rs
use crate::flamegraph::VizData;

const RESET: &str = "\x1b[0m";
const BOLD: &str = "\x1b[1m";
const DIM: &str = "\x1b[2m";
const FULL: &str = "█";

fn color(cat: &str) -> &'static str {
    match cat {
        "thinking" => "\x1b[91m",
        "text" => "\x1b[94m",
        "read" => "\x1b[92m",
        "write" => "\x1b[93m",
        "bash" => "\x1b[95m",
        "search" => "\x1b[32m",
        "agent" => "\x1b[31m",
        "other" => "\x1b[96m",
        "cache" => "\x1b[90m",
        "fresh" => "\x1b[37m",
        _ => "\x1b[96m",
    }
}

fn bar(value: usize, max_val: usize, cat: &str, width: usize) -> String {
    if max_val == 0 { return String::new(); }
    let filled = (value as f64 / max_val as f64 * width as f64).ceil() as usize;
    let filled = filled.min(width);
    format!("{}{}{}", color(cat), FULL.repeat(filled), RESET)
}

fn fmt_tokens(n: usize) -> String {
    if n >= 1_000_000 { format!("{:.1}M", n as f64 / 1_000_000.0) }
    else if n >= 1_000 { format!("{:.1}K", n as f64 / 1_000.0) }
    else { format!("{}", n) }
}

fn pct(n: usize, total: usize) -> String {
    if total == 0 { "0%".to_string() }
    else { format!("{:.1}%", n as f64 / total as f64 * 100.0) }
}

fn score_color(s: f64) -> &'static str {
    if s >= 80.0 { "\x1b[92m" }
    else if s >= 60.0 { "\x1b[93m" }
    else { "\x1b[91m" }
}

pub fn render_dashboard(viz: &VizData) {
    let w = 80;
    println!("{BOLD}{}{RESET}", "═".repeat(w));
    println!("{BOLD}  Token Flamegraph{RESET}");
    if !viz.model.is_empty() {
        println!("  {DIM}Model: {}{RESET}", viz.model);
    }
    println!("{BOLD}{}{RESET}", "═".repeat(w));

    let t = &viz.totals;
    let avg_score: f64 = if viz.turns.is_empty() { 0.0 } else {
        viz.turns.iter().map(|t| t.efficiency_score).sum::<f64>() / viz.turns.len() as f64
    };
    let cache_ratio = if t.input > 0 { t.cache as f64 / t.input as f64 * 100.0 } else { 0.0 };

    // Cost estimate (Opus pricing rough)
    let input_cost = t.input as f64 / 1_000_000.0 * 15.0;
    let cached_cost = t.cache as f64 / 1_000_000.0 * 1.5;
    let fresh_cost = (t.input - t.cache) as f64 / 1_000_000.0 * 15.0;
    let output_cost = t.output as f64 / 1_000_000.0 * 75.0;
    let total_cost = cached_cost + fresh_cost + output_cost;
    let saved = input_cost - (cached_cost + fresh_cost);

    println!();
    println!("  {BOLD}Output:{RESET} {:<12} {BOLD}Thinking:{RESET} {:<12} {BOLD}Tool Calls:{RESET} {:<6} {BOLD}Turns:{RESET} {:<4} {BOLD}Score:{RESET} {}⚡{:.0}{RESET}",
        fmt_tokens(t.output), fmt_tokens(t.thinking), t.tool_calls, viz.turns.len(), score_color(avg_score), avg_score);
    println!("  {BOLD}Input:{RESET}  {:<12} {BOLD}Cached:{RESET} {} ({:.1}%)    {BOLD}Cost:{RESET} ${:.2} (saved ${:.2} from cache)",
        fmt_tokens(t.input), fmt_tokens(t.cache), cache_ratio, total_cost, saved);

    // Timeline
    println!();
    println!("  {DIM}{}{RESET}", "─".repeat(w - 4));
    println!("  {BOLD}TIMELINE — Output Tokens per Turn{RESET}");
    println!("  {DIM}{}{RESET}", "─".repeat(w - 4));

    let max_output = viz.turns.iter().map(|t| t.output_tokens).max().unwrap_or(1);

    for turn in &viz.turns {
        let sc = turn.efficiency_score;
        let sc_col = score_color(sc);

        // Main bar
        let mut segments = Vec::new();
        if turn.thinking > 0 { segments.push(bar(turn.thinking, max_output, "thinking", 40)); }
        for tool in &turn.tools {
            segments.push(bar(tool.total_tokens, max_output, &tool.category, 40));
        }
        if turn.text > 0 { segments.push(bar(turn.text, max_output, "text", 40)); }

        let bar_str = segments.join("");
        println!();
        println!("  {BOLD}Turn {:2}{RESET} │{bar_str}│  {:>6} tok  {sc_col}⚡{:>4.0}{RESET}",
            turn.index, fmt_tokens(turn.output_tokens), sc);

        // Tool details
        if !turn.tools.is_empty() {
            let names: Vec<String> = turn.tools.iter().take(6).map(|t| {
                let c = color(&t.category);
                let short: String = t.name.chars().take(28).collect();
                format!("{c}{short}{RESET}")
            }).collect();
            let more = if turn.tools.len() > 6 { format!(" {DIM}+{} more{RESET}", turn.tools.len() - 6) } else { String::new() };
            println!("         │ {}{more}", names.join(", "));
        }
    }

    // Wall time (if available)
    if viz.turns.iter().any(|t| t.duration_ms > 0) {
        println!();
        println!("  {DIM}{}{RESET}", "─".repeat(w - 4));
        println!("  {BOLD}WALL TIME PER TURN{RESET}");
        println!("  {DIM}{}{RESET}", "─".repeat(w - 4));

        let max_dur = viz.turns.iter().map(|t| t.duration_ms).max().unwrap_or(1);
        let total_wall: u64 = viz.turns.iter().map(|t| t.duration_ms).sum();
        let total_api: u64 = viz.turns.iter().map(|t| t.duration_api_ms).sum();

        for turn in &viz.turns {
            if turn.duration_ms == 0 { continue; }
            let api_bar = bar(turn.duration_api_ms as usize, max_dur as usize, "thinking", 30);
            let overhead_bar = bar((turn.duration_ms - turn.duration_api_ms) as usize, max_dur as usize, "bash", 10);
            let model_pct = turn.duration_api_ms * 100 / turn.duration_ms.max(1);
            println!("  T{:<3} {api_bar}{overhead_bar} {:>6.1}s ({}% model)",
                turn.index, turn.duration_ms as f64 / 1000.0, model_pct);
        }
        println!("  {DIM}Total: {:.1}s wall / {:.1}s model / {:.1}s overhead{RESET}",
            total_wall as f64 / 1000.0, total_api as f64 / 1000.0, (total_wall - total_api) as f64 / 1000.0);
    }

    // Output breakdown
    println!();
    println!("  {DIM}{}{RESET}", "─".repeat(w - 4));
    println!("  {BOLD}OUTPUT TOKEN BREAKDOWN{RESET}");
    println!("  {DIM}{}{RESET}", "─".repeat(w - 4));

    let mut categories: Vec<(&str, usize, &str)> = vec![
        ("Thinking", t.thinking, "thinking"),
        ("Text output", t.text, "text"),
    ];
    let cat_names: std::collections::HashMap<&str, &str> = [
        ("read", "File reads"), ("write", "File writes"), ("bash", "Shell commands"),
        ("search", "Search"), ("agent", "Sub-agents"), ("other", "Other tools"),
    ].into();
    let mut cat_list: Vec<_> = viz.category_tokens.iter().collect();
    cat_list.sort_by(|a, b| b.1.cmp(a.1));
    for (cat, tokens) in &cat_list {
        categories.push((cat_names.get(cat.as_str()).unwrap_or(&cat.as_str()), **tokens, cat));
    }
    let max_cat = categories.iter().map(|(_, t, _)| *t).max().unwrap_or(1);
    let total_cat: usize = categories.iter().map(|(_, t, _)| *t).sum();
    for (name, tokens, cat) in &categories {
        let b = bar(*tokens, max_cat, cat, 35);
        println!("  {:<16} {b} {:>7} ({})", name, fmt_tokens(*tokens), pct(*tokens, total_cat));
    }

    // Top tools
    if !viz.top_tools.is_empty() {
        println!();
        println!("  {DIM}{}{RESET}", "─".repeat(w - 4));
        println!("  {BOLD}TOP TOKEN CONSUMERS{RESET}");
        println!("  {DIM}{}{RESET}", "─".repeat(w - 4));

        let max_tool = viz.top_tools[0].tokens;
        for tool in viz.top_tools.iter().take(10) {
            let cat = if tool.name.starts_with("Read") { "read" }
                else if tool.name.starts_with("Edit") || tool.name.starts_with("Write") { "write" }
                else if tool.name.starts_with("Bash") { "bash" }
                else if tool.name.starts_with("Grep") || tool.name.starts_with("Glob") { "search" }
                else if tool.name.starts_with("Agent") { "agent" }
                else { "other" };
            let b = bar(tool.tokens, max_tool, cat, 30);
            let short: String = tool.name.chars().take(42).collect();
            println!("  {:<44} {b} {:>6}", short, fmt_tokens(tool.tokens));
        }
    }

    println!();
    println!("{BOLD}{}{RESET}", "═".repeat(w));

    // Legend
    let legend = [("thinking", "Thinking"), ("text", "Text"), ("write", "Write"),
        ("read", "Read"), ("bash", "Shell"), ("search", "Search"), ("agent", "Agent")];
    let leg: Vec<String> = legend.iter().map(|(cat, name)| format!("{}{FULL}{FULL}{RESET} {name}", color(cat))).collect();
    println!("  {}", leg.join("  "));
    println!();
}
